Created docker tutorial for redis

DoruM opened the pull request:
A few comments on this pull request. 

- curl uses Chrome user agent (instead of curl user agent). This is done in order to get over the firewall.
- curl uses --insecure. This was done for the same reason. 
- I have allowed origin * in the gateway.
- Changed to ws from wss
- Added volume mapping for the drives in redis (Thus they can be checked directly from the host, or from boot2docker).
- This example uses the JMS client. If a redis only client is needed, we have to create it first. As such a new issue should be created.

dpwspoon commented on redis/redis/docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+  redis:
+    build: broker
+    ports:
+      - "6379:6379"
```
I don't think this port needs to be exposed, correct?  My thoughts are that production deployments would not need this

DoruM replied:
To be honest, this depends on the design of the system. Assuming that gateway is the only client of the redis, your comment is correct. I assumed that different clients might be required to connect to redis, other than the gateway. I will remove it, since I think this is not the case.

dpwspoon commented on redis/redis/docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+    volumes:
+     - /var/log/tut-redis/redis:/data/log/
+networks:
+  tut-redis:
```
My philosophy on creating this is to keep everything bare minimal unless absolutely needed, with that stated I don't see a reason for the network config and assigning ip addresses, was this added for a specific reason?

DoruM replied:
First of all, I think this is best practices in docker to assign IP addresses.

Aside from that, there was a practical reason for that. I run docker in Windows 7 (boot2docker).

As such, to connect to the actual server I need to create a tunnel in VirtualBox to the machine, and connect to THAT machine, (from Chrome from the Windows 7 host). IF I wouldn't create a static IP, a new IP would be created every time I would tear it down and I would need to enter into VirtualBox GUI and change the tunnel every time.

BTW, related to this, I think more information should be added on the README as to how to run on boot2docker.

dpwspoon commented on redis/redis/gateway/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+
+# Add and Build Javascript Client tutorials to web directory so we can test it from a browser,
+# TODO pull in built tutorials when they are released
+
```
This TODO can be resolved now that the tutorials are build and can be pulled directly from https://github.com/kaazing/javascript.client.tutorials/releases.  This removes the need to install node and should simplify this

dpwspoon commented:
I would advocate not adding the volumes for the same reason I mentioned here: `https://github.com/nginxinc/docker-nginx/issues/19`, i.e.

> I discovered this when I was trying to learn docker and I noticed I was unintentionally orphaning docker volumes.

unless there is an very compelling reason to do so.  IMHO I don't think the logs use case fits that reason which is why I didn't add it to the base Dockerfile for gateway and enterprise gateway.  I've seen other base images that do add logs to the volume but I believe people are trending away from doing that.  For example [Tomcat](https://hub.docker.com/_/tomcat/) does not have it.

dpwspoon commented on redis/redis/broker/Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM redis:latest
+
+COPY redis.conf /usr/local/etc/redis/redis.conf
```
Was there something that needed to be customized in the redis config?  If so please add that to the readme, as the file is too long for end users to do a manual diff.  If not, please remove the config (and perhaps this dockerfile via referencing the redis:latest directly from docker compose)

DoruM replied:
Yes, there was a reason for it.

The main reason was that I wanted to see the redis logs. Default writes to stdout. This is not ok in my opinion, if you want to examine these logs later.

dpwspoon commented:
>
I have allowed origin * in the gateway.
Changed to ws from wss


What is the reasoning here?  In general I'd like the examples / tutorials in the repository to reflect and build on each other.  If these stopped you from making progress perhaps we should the other tutorials to not use wss and allow *.

dpwspoon commented:
Thanks @DoruM

DoruM commented:
Hi David, 

Regarding https://github.com/kaazing/docker.tutorials/pull/5#issuecomment-229714794 , I feel that especially with tutorials logs should be in a volume. 

The purpose of these tutorials is for the users to modify them to fit their needs. If the gateway (or redis) will crash as a result of these changes, the logs will be forever gone. To avoid this, we place them in a volume. 

If you really need this change, I can implement it, of course.

Best regards,
Doru


@dpwspoon

dpwspoon commented:
There are clearly different ways people can manage logging on containers / with docker [1](https://www.loggly.com/blog/top-5-docker-logging-methods-to-fit-your-container-deployment-strategy/) and [2](https://www.quora.com/What-is-a-best-practice-for-log-management-for-dockerized-apps).  My view has always been that outputting to std err / std out and using docker [log](https://docs.docker.com/engine/reference/commandline/logs/) and [drivers](https://docs.docker.com/engine/admin/logging/overview/) functionality is/will be the best.  

Doing a web search will give a variety of different answers whether to do "application logging" via volumes or "container logging" with the debate going several different ways.  

IMHO container logging via docker drivers is superior as it gives a unified approach to logging all services.  Based on that, that is what we would recommend customers do in production and I would think all our tutorials shipping gateway should point them to that direction.  And thats why we have done that in our base images and in the other tutorials.  I'm happy to debate which way is better (via slack), but for consistency please follow the other tutorials here.  If we choose to change them all in the future I'm fine with that.

-- Side note.

Additionally, I've been trying to keep the tutorials bare bones as I suspect people might get scared away at complexity.  Thus minimizing content and steps to getting a working example I think is ideal.  Though chief demo officer @nemigaservices might have an opinion on that. :-)

DoruM commented:
Very well, I'll remove volumes and fixed IPs.

DoruM commented:
Hi David, 

I have modified this release to use proper "allow origin". I have also modified to use released version of javascript JMS client, instead of compiling it using node.js.

I was however unable to make it work using wss. 

-----------------------------------------
Aside from this, I have created another [PR](https://github.com/kaazing/docker.tutorials/pull/6) which has implemented all of your ideas. You can either merge [that one](https://github.com/kaazing/docker.tutorials/pull/6) or this one, but not both.

Which is merged, will depend upon what are actually the best practices. This pull request contains logging in volumes, fixed IPs,etc.  [That pull request](https://github.com/kaazing/docker.tutorials/pull/6) contains logging to stdout and no fixed IPs. 

Best regards,
Doru

@dpwspoon
